src/main.rs

```diff
@@ -6,44 +6,33 @@ extern crate log;
 extern crate alloc;
 extern crate axstd;
 
+mod loader;
 mod mm;
 mod syscall;
 mod task;
-
-use alloc::sync::Arc;
+use alloc::{sync::Arc, vec::Vec};
 
 use axhal::arch::UspaceContext;
 use axsync::Mutex;
 
-const USER_STACK_SIZE: usize = 4096;
+const USER_STACK_SIZE: usize = 0x10000;
 const KERNEL_STACK_SIZE: usize = 0x40000; // 256 KiB
 
-fn app_main(arg0: usize) {
-    unsafe {
-        core::arch::asm!(
-            "2:",
-            "int3",
-            "mov rax, r12",
-            "push rax",
-            "syscall",
-            "add r12, 1",
-            "jmp 2b",
-            in("r12") arg0,
-            in("rdi") 2,
-            in("rsi") 3,
-            in("rdx") 3,
-            options(nostack, nomem)
-        )
-    }
-}
-
 #[no_mangle]
 fn main() {
-    let (entry_vaddr, ustack_top, uspace) = mm::load_user_app(app_main).unwrap();
-    let user_task = task::spawn_user_task(
-        Arc::new(Mutex::new(uspace)),
-        UspaceContext::new(entry_vaddr.into(), ustack_top, 2333),
-    );
-    let exit_code = user_task.join();
-    info!("User task exited with code: {:?}", exit_code);
+    loader::list_apps();
+    let testcases: Vec<&'static str> = option_env!("AX_TESTCASES_LIST")
+        .unwrap_or_else(|| "Please specify the testcases list by making user_apps")
+        .split(',')
+        .filter(|&x| !x.is_empty())
+        .collect();
```

Comment:
No need to collect into `Vec`? You can iterate over the value returned by `filter`.